packages/liveblocks-client/src/shallow.test.ts

```diff
@@ -0,0 +1,158 @@
+
+describe("shallow", () => {
+  it("scalar values", () => {
+    expect(shallow(0, 0)).toBe(true);
+    expect(shallow("", "")).toBe(true);
+    expect(shallow("hi", "hi")).toBe(true);
+    expect(shallow(false, false)).toBe(true);
+    expect(shallow(true, true)).toBe(true);
+
+    expect(shallow(0, 1)).toBe(false);
+    expect(shallow(false, true)).toBe(false);
+    expect(shallow(true, false)).toBe(false);
+
+    // Weird exceptions
+    expect(shallow(NaN, NaN)).toBe(true);
+    expect(shallow(-0, +0)).toBe(false);
+  });
+
+  it("scalar values wrapped in list", () => {
+    expect(shallow([0], [0])).toBe(true);
+    expect(shallow([""], [""])).toBe(true);
+    expect(shallow(["hi"], ["hi"])).toBe(true);
+    expect(shallow([false], [false])).toBe(true);
+    expect(shallow([true], [true])).toBe(true);
+
+    expect(shallow([0], [1])).toBe(false);
+    expect(shallow([false], [true])).toBe(false);
+    expect(shallow([true], [false])).toBe(false);
+
+    // Weird exceptions   ],[
+    expect(shallow([NaN], [NaN])).toBe(true);
+    expect(shallow([-0], [+0])).toBe(false);
+  });
+
+  it("scalar values wrapped in objs", () => {
+    expect(shallow({ k: 0 }, { k: 0 })).toBe(true);
+    expect(shallow({ k: "" }, { k: "" })).toBe(true);
+    expect(shallow({ k: "hi" }, { k: "hi" })).toBe(true);
+    expect(shallow({ k: false }, { k: false })).toBe(true);
+    expect(shallow({ k: true }, { k: true })).toBe(true);
+
+    expect(shallow({ k: 0 }, { k: 1 })).toBe(false);
+    expect(shallow({ k: false }, { k: true })).toBe(false);
+    expect(shallow({ k: true }, { k: false })).toBe(false);
+
+    // Weird exceptions
+    expect(shallow({ k: NaN }, { k: NaN })).toBe(true);
+    expect(shallow({ k: -0 }, { k: +0 })).toBe(false);
+  });
+});
+
+describe("shallow (properties)", () => {
+  it("all equal scalars are shallowly equal", () => {
+    fc.assert(
+      fc.property(
+        // Inputs
+        fc.clone(scalar(), 2),
+
+        // Unit test
+        ([scalar1, scalar2]) => {
+          expect(shallow(scalar1, scalar2)).toBe(true);
+
+          // If two values are shallowly equal, then those values wrapped in an
+          // array (one level) should _also_ be considered shallowly equal
+          expect(shallow([scalar1], [scalar2])).toBe(true);
+          expect(shallow([scalar1, scalar1], [scalar2, scalar2])).toBe(true);
+
+          // ...but wrapping twice is _never_ going to be equal
+          expect(shallow([[scalar1]], [[scalar2]])).toBe(false);
+
+          // Ditto for objects
+          expect(shallow({ a: scalar1 }, { a: scalar2 })).toBe(true);
+          expect(
+            shallow({ a: scalar1, b: scalar1 }, { a: scalar2, b: scalar2 })
+          ).toBe(true);
+
+          // ...but nesting twice is _never_ going to be equal
+          expect(shallow({ a: { b: scalar1 } }, { a: { b: scalar2 } })).toBe(
+            false
+          );
+        }
+      )
+    );
+  });
```

Comment:
Specifically in this test:

- On L65, we're generating a random scalar, and "clone" it to duplicate it, so if it generated `'hi'`, we'll now get `['hi', 'hi']`
- On L68, the actual unit test receives the two scalars and unpacks them
- On L69, we verify that `shallow()` will indeed consider those values equal. Due to the way these inputs are generated, we _know_ these are scalars — so never will be complex arrays/objects, and they should be the same.
- On L71-74, we wrap these values in an array, and verify that those must also always be equal according to the rules
- On L76-77, we also verify that double-wrapping will _always_ be **un**equal!
- On L79-83, we do the same, but for objects, not arrays

**Note that these tests, because random every time, will over time cover more and more possible inputs, and over time increase the confidence that this implementation is free of bugs.**